internal/catalog/internal/types/acl_hooks.go

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package types
+
+import (
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/internal/resource"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+func aclReadHookResourceWithWorkloadSelector(authorizer acl.Authorizer, authzContext *acl.AuthorizerContext, id *pbresource.ID, _ *pbresource.Resource) error {
+	return authorizer.ToAllowAuthorizer().ServiceReadAllowed(id.GetName(), authzContext)
+}
+
+func aclWriteHookResourceWithWorkloadSelector[T WorkloadSelecting](authorizer acl.Authorizer, authzContext *acl.AuthorizerContext, res *pbresource.Resource) error {
+	if res == nil {
+		return resource.ErrNeedResource
+	}
+
+	decodedService, err := resource.Decode[T](res)
+	if err != nil {
+		return resource.ErrNeedResource
+	}
+
+	// First check service:write on the name.
+	err = authorizer.ToAllowAuthorizer().ServiceWriteAllowed(res.GetId().GetName(), authzContext)
+	if err != nil {
+		return err
+	}
+
+	// Then also check whether we're allowed to select a service.
+	for _, name := range decodedService.GetData().GetWorkloads().GetNames() {
+		err = authorizer.ToAllowAuthorizer().ServiceReadAllowed(name, authzContext)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, prefix := range decodedService.GetData().GetWorkloads().GetPrefixes() {
+		err = authorizer.ToAllowAuthorizer().ServiceReadAllowed(prefix, authzContext)
```

Comment:
Is this safe? I'm imagining a case where a workload selector is selecting all workloads with a prefix of `api` and they have `service_prefix "api1" { policy = "read" }`. I believe this check would fail becaue they don't have service read or service prefix read on `api`.

Reply:
The current behavior is to deny this write (not sure if that's what you meant by this check failing). 

I think this is correct because if you have a workload selector `Workloads = { Prefixes = ["api"] }`, your service, for example, can select any workloads that start with `api` (`api1`, `api2` etc). But your policy only allows you to write a service selecting workloads that start with `api1`. And so we should deny this request because otherwise you can select more workloads than what you're allowed to.

Let me know if I misunderstood you though!

Reply:
We talked about this in private and agreed to merging this as is and following up next week.

Reply:
I think to properly solve this we need to have a different way to authorize prefixes which will involve more changes to policy authorizer. I think it makes sense to do in a separate PR.